### PR TITLE
Rework ASTDumper

### PR DESCRIPTION
**Pull request by beccadax**
This PR refactors the ASTDumper to make it more structured, less mistake-prone, and more amenable to future changes. For example:

```cpp
  // Before:
  void visitUnresolvedDotExpr(UnresolvedDotExpr *E) {
    printCommon(E, "unresolved_dot_expr")
      << " field '" << E->getName() << "'";
    PrintWithColorRAII(OS, ExprModifierColor)
      << " function_ref=" << getFunctionRefKindStr(E->getFunctionRefKind());
    if (E->getBase()) {
      OS << '\n';
      printRec(E->getBase());
    }
    PrintWithColorRAII(OS, ParenthesisColor) << ')';
  }

  // After:
  void visitUnresolvedDotExpr(UnresolvedDotExpr *E, StringRef label) {
    printCommon(E, "unresolved_dot_expr", label);

    printFieldQuoted(E->getName(), "field");
    printField(E->getFunctionRefKind(), "function_ref", ExprModifierColor);

    if (E->getBase()) {
      printRec(E->getBase());
    }

    printFoot();
  }
```

* Values are printed through calls to base class methods, rather than direct access to the underlying `raw_ostream`.
    * These methods tend to reduce the chances of bugs like missing/extra spaces or newlines, too much/too little indentation, etc.
    * More values are quoted, and unprintable/non-ASCII characters in quoted values are escaped before printing.
* Infrastructure to label child nodes now exists.
    * Some weird breaks from the normal "style", like `PatternBindingDecl`'s original and processed initializers, have been brought into line.
* Some types that previously used ad-hoc dumping functions, like conformances and substitution maps, are now structured similarly to the dumper classes.
* I've fixed the odd dumping bug along the way. For example, distributed actors were only marked `actor`, not `distributed actor`.

This PR doesn't change the overall style of AST dumps; they're still pseudo-S-expressions. But the logic that implements this style is now isolated into a relatively small base class, making it feasible to introduce e.g. JSON dumping in the future.

### CODE REVIEW

**Comment by beccadax** [comment]
@swift-ci please smoke test

File: test/expr/capture/generic_params.swift [hunk 1]
```diff
@@ -2,33 +2,33 @@
...
   // Special case -- closure captures outer generic parameter, but it does not
   // appear as the type of any expression
-  // CHECK: closure_expr type='() -> ()' {{.*}} discriminator=2 captures=(<generic> x<direct>)
+  // CHECK: closure_expr type="() -> ()" {{.*}} discriminator=2 captures=(<generic> x<direct>)
   _ = { if x is T {} }
 
   // Nested generic functions always capture outer generic parameters, even if
-  // they're not mentioned in the function body
-  // CHECK: func_decl{{.*}}"innerGeneric(u:)" <U> interface type='<T, U> (u: U) -> ()' {{.*}} captures=(<generic> )
+  // they"re not mentioned in the function body
```
Comment (bnbarham) [inline test/expr/capture/generic_params.swift:22]: This is the only bad replacement I noticed, but I only took a super quick scan (and doesn't really matter)

Comment (beccadax) [reply test/expr/capture/generic_params.swift]: Fixed it. Still perfectly balanced!

**Comment by beccadax** [comment]
Changes:
* Corrected some commit messages
* Fixed one comment change where an apostrophe was incorrectly changed to a double-quote

**Comment by beccadax** [comment]
* More commit message corrections

**Comment by beccadax** [comment]
The test jobs from before I started force-pushing are:
* https://ci.swift.org/job/swift-PR-macos-smoke-test/9094/
* https://ci.swift.org/job/swift-PR-Linux-smoke-test/8725/
* https://ci-external.swift.org/job/swift-PR-windows/16511/

**Comment by beccadax** [comment]
@swift-ci please test

**Review by ktoso — APPROVED** [review]
AFAICS nice improvements 👍

File: lib/AST/ASTDumper.cpp [hunk 2]
```diff
@@ -114,27 +135,62 @@ namespace {
...
+    return base_os.has_colors();
+  }
+
+  virtual void enable_colors(bool enable) override {
+    base_os.enable_colors(enable);
+  }
+
+private:
+  virtual void write_impl(const char *Ptr, size_t Size) override {
+    base_os.write_escaped(StringRef(Ptr, Size), /*UseHexEscapes=*/true);
```
Comment (AnthonyLatsis) [inline lib/AST/ASTDumper.cpp:182]: FYI, this made the tau in generic parameter names print as escape sequences:
```
(primary_archetype_type address=true name="\xCF\x84_0_0"
  (interface_type=generic_type_param_type depth=0 index=0))
```

Comment (beccadax) [reply lib/AST/ASTDumper.cpp:182]: True. Using `write_escaped` happened to be very convenient, but I wouldn't be averse to a patch that excepted the tau character, or some larger set of known-printable characters (as long as it'll still escape various newlines, quotes, control characters, etc.).

**Pull request merged**